### PR TITLE
Make possible screenshots on hybrid sdks (react-native)

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :scroll: Description
To make screenshots work on RN we need to set the current activity for the ScreenshotEventProcessor when initializing the Android SDK otherwise [it will be null](https://github.com/getsentry/sentry-react-native/issues/2362#issuecomment-1193763553).

We will be sending the event with a screenshot therefore we need to ensure another screenshot won't be added by the Android SDK.

## :bulb: Motivation and Context
- https://github.com/getsentry/team-mobile/issues/11
- https://github.com/getsentry/sentry-react-native/pull/2610

## :pencil: Checklist

- [x] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

## :crystal_ball: Next steps
#skip-changelog

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:60]: L: any reason to move the code?

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:65]: H: `EnvelopeFileObserver` uses the `CachedEnvelopeHint` hint so this won't work, we need to extend the current hint type for Hybrid SDKs most likely.
The goal here is that at least for screenshots, screenshots should not be taken (again) for Hybrid SDKs.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (krystofwoldrich) [reply sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:60]: I don't know how is the lint set in this project, but I saw everywhere the public methods are before private.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (krystofwoldrich) [reply sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:65]: Okay got it, that was my goal, I've just misunderstood the function.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:57]:
```suggestion
  public static @NotNull ScreenshotEventProcessor getInstance(
```

H: Always use the nullability annotations, better Kotlin interoperability.
M: Stick to the convention of `getInstance` naming.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:63]: H: if `instance` isn't null, return itself.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:37]: H: if `createInstance` was not called, this returns null, so it should be nullable or we should not use a singleton pattern

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:66]: Similar to https://github.com/getsentry/sentry-java/pull/2360/files#r1023634311

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:133]: M: I'd move the `getIsFromHybridSdk` later otherwise Native events pay the price every time even if the integration isn't installed nor its an error event.

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/util/HintUtils.java:22]: M: The class is already `@ApiStatus.Internal`, no need to add to every method.

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/util/HintUtils.java:25]: L: Other Hybrid SDKs would have the same issue I suspect, but we can expand this later.

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/util/HintUtils.java:36]:
```suggestion
  public static boolean isFromHybridSdk(Hint hint) {
```

**Comment by marandaneto** [comment]
Missing a few things, nullability annotations, naming convention, and tests.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:40]: H: You should always access `CurrentActivityHolder.getInstance()` directly.

File: sentry-android-core/src/main/java/io/sentry/android/core/CurrentActivityHolder.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/CurrentActivityHolder.java:22]: H: Check how other `getInstance()` is done, this implementation is not multithreaded safe.

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ActivityUtils.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ActivityUtils.java:14]: H: The `buildInfoProvider` has been removed and now this cannot be unit tested.

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java:16]:
```suggestion
    this("Sentry");
```

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java:13]:
```suggestion
  private final @NotNull String tag;
```

M: double check all the missing nullability annotations

File: sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/AndroidLogger.java:19]:
```suggestion
  public AndroidLogger(final @NotNull String tag) {
```

M: Always make fields `final` when possible, there are more cases in this PR

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:66]:
```suggestion
    final Activity activity = CurrentActivityHolder.getInstance().getActivity()
    if (activity == null
```

M: Now you can reuse the `activity` object which lives only in the scope of this method.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:81]:
```suggestion
    hint.set(ANDROID_ACTIVITY, activity);
```

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java:20]: L: I guess we can do the `activity` check before and make it `NotNull` here.

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java:69]:
```suggestion
  private static boolean isActivityValid(
```

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/util/HintUtils.java:29]: M: This can be private or inlined with the other `setIsFromHybridSdk` method, I don't see it being called from somewhere else.

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (krystofwoldrich) [reply sentry/src/main/java/io/sentry/util/HintUtils.java]: I've changed it a bit because we can consider all the SDKs based on JS, Dotnet, or Dart them hybrid then the check can be more universal.

File: sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/internal/util/ScreenshotUtils.java:45]: M: Likely an improvement I forgot to do before:
`ByteArrayOutputStream` can be created within a try with resources so it auto closes the stream, right now if it fails for some reason, the stream is kept open.
Example:
`try (final ByteArrayOutputStream stream = new ByteArrayOutputStream();) {}`

**Comment by marandaneto** [comment]
Left the last comment but all is good on my side.
I will let the maintainers do the final approval.

**Review by marandaneto — APPROVED** [review]

**Comment by romtsn** [comment]
it's just flaky

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (romtsn) [inline sentry/src/main/java/io/sentry/util/HintUtils.java:23]: `l`: should these 2 new methods be also marked as `@Internal`?

File: sentry-android-core/src/main/java/io/sentry/android/core/CurrentActivityHolder.java
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/CurrentActivityHolder.java:14]: `m`: can we also add a private constructor here so it's not possible to instantiate the `CurrentActivityHolder` and we are always forced to use the singleton instance?
```suggestion

    private CurrentActivityHolder() { }
```

**Review by romtsn — APPROVED** [review]
lgtm 👍 can you also please add a changelog entry - I understand that it's more-or-less an internal change, but if something goes wrong and this PR causes bugs, we can quickly look at the past releases and see if there was something suspicious in the changelog.

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/util/HintUtils.java:23]: The class is already internal.

**Comment by marandaneto** [comment]
> lgtm 👍 can you also please add a changelog entry - I understand that it's more-or-less an internal change, but if something goes wrong and this PR causes bugs, we can quickly look at the past releases and see if there was something suspicious in the changelog.

Changelog is meant for user-facing changes, I'd not add it but your call :)

File: sentry/src/main/java/io/sentry/util/HintUtils.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/util/HintUtils.java:23]: then we should probably remove it from the other methods to keep it aligned

**Pull request merged**
